File: ios/socketx_client/Sources/socketx_client/SocketXClientPlugin.swift
// The MIT License (MIT)
//
// Copyright (c) Eclypses, Inc.
//
// All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import Flutter
import UIKit
import SocketXClient

public class SocketXClientPlugin: NSObject, FlutterPlugin {
    
    // MARK: - Class Variables
    private var methodChannel: FlutterMethodChannel?
    private static let channelName = "socketx_client"
    private var socketXClient: SocketXClient?
    private var urlSession: URLSession?
    
    // MARK: - Plugin Registration
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = SocketXClientPlugin()
        let methodChannel = FlutterMethodChannel(name: channelName, binaryMessenger: registrar.messenger())
        instance.methodChannel = methodChannel
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
    }
    
    // MARK: - Method Call Handler
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
            
        case "connect":
            handleConnect(call: call, result: result)
            
        case "disconnect":
            handleDisconnect(result: result)
            
        case "sendText":
            handleSendText(call: call, result: result)
            
        case "sendBinary":
            handleSendBinary(call: call, result: result)

        case "enableFileLogging":
            handleEnableFileLogging(call: call, result: result)

        case "readLogFile":
            handleReadLogFile(result: result)

        case "clearLogFile":
            SocketXClient.clearLogFile()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Diagnostics

    private func handleEnableFileLogging(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let enabled = args["enabled"] as? Bool else {
            result(FlutterError(code: "INVALID_ARGUMENTS",
                               message: "'enabled' is required",
                               details: nil))
            return
        }
        SocketXClient.enableFileLogging(enabled)
        result(nil)
    }

    private func handleReadLogFile(result: @escaping FlutterResult) {
        do {
            result(try SocketXClient.readLogFile())
        } catch {
            result(FlutterError(code: "READ_LOG_ERROR",
                               message: error.localizedDescription,
                               details: nil))
        }
    }
    
    // MARK: - Connection Methods
    
    private func handleConnect(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let urlString = args["url"] as? String,
              let url = URL(string: urlString) else {
            result(FlutterError(code: "INVALID_ARGUMENTS",
                               message: "URL is required",
                               details: nil))
            return
        }
        
        // Disconnect existing connection if any
        socketXClient?.disconnect()
        
        // Create URL session configuration
        let configuration = URLSessionConfiguration.default
        if let headers = args["headers"] as? [String: String] {
            configuration.httpAdditionalHeaders = headers
        }
        
        // Create URL session and WebSocket task
        urlSession = URLSession(configuration: configuration)
        guard let task = urlSession?.webSocketTask(with: url) else {
            result(FlutterError(code: "CONNECTION_ERROR",
                               message: "Failed to create WebSocket task",
                               details: nil))
            return
        }
        
        do {
            // Initialize SocketXClient with the WebSocket task
            socketXClient = try SocketXClient(task: task)
            setupCallbacks()
            socketXClient?.connect()
            result(nil)
        } catch {
            sendError(type: "internal", reason: error.localizedDescription)
            result(FlutterError(code: "INITIALIZATION_ERROR",
                               message: error.localizedDescription,
                               details: nil))
        }
    }
    
    private func handleDisconnect(result: @escaping FlutterResult) {
        socketXClient?.disconnect()
        socketXClient = nil
        urlSession?.invalidateAndCancel()
        urlSession = nil
        result(nil)
    }
    
    // MARK: - Send Methods
    
    private func handleSendText(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let text = args["text"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS",
                               message: "Text is required",
                               details: nil))
            return
        }
        
        guard let client = socketXClient else {
            result(FlutterError(code: "NOT_CONNECTED",
                               message: "WebSocket is not connected",
                               details: nil))
            return
        }
        
        client.send(text: text)
        result(nil)
    }
    
    private func handleSendBinary(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let data = args["data"] as? FlutterStandardTypedData else {
            result(FlutterError(code: "INVALID_ARGUMENTS",
                               message: "Binary data is required",
                               details: nil))
            return
        }
        
        guard let client = socketXClient else {
            result(FlutterError(code: "NOT_CONNECTED",
                               message: "WebSocket is not connected",
                               details: nil))
            return
        }
        
        client.send(binary: data.data)
        result(nil)
    }
    
    // MARK: - Callback Setup
    
    private func setupCallbacks() {
        socketXClient?.onConnected = { [weak self] in
            self?.invokeMethod("onConnected", arguments: nil)
        }
        
        socketXClient?.onMessageReceived = { [weak self] text in
            self?.invokeMethod("onMessage", arguments: text)
        }
        
        socketXClient?.onBinaryReceived = { [weak self] data in
            self?.invokeMethod("onBinaryMessage", arguments: FlutterStandardTypedData(bytes: data))
        }
        
        socketXClient?.onError = { [weak self] error in
            self?.handleSocketXError(error)
        }
    }
    
    // MARK: - Error Handling
    
    private func handleSocketXError(_ error: SocketXError) {
        let errorType: String
        let reason: String
        
        switch error {
        case .networkError(let r):
            errorType = "network"
            reason = r
        case .transportError(let r):
            errorType = "transport"
            reason = r
        case .codecError(let r):
            errorType = "codec"
            reason = r
        case .handshakeError(let r):
            errorType = "handshake"
            reason = r
        case .proxyError(let r):
            errorType = "proxy"
            reason = r
        case .internalError(let r):
            errorType = "internal"
            reason = r
        case .unknown(let r):
            errorType = "unknown"
            reason = r
        }
        
        sendError(type: errorType, reason: reason)
    }
    
    private func sendError(type: String, reason: String) {
        let errorMap: [String: Any] = [
            "type": type,
            "reason": reason
        ]
        invokeMethod("onError", arguments: errorMap)
    }
    
    // MARK: - Helper Methods
    
    private func invokeMethod(_ method: String, arguments: Any?) {
        DispatchQueue.main.async { [weak self] in
            self?.methodChannel?.invokeMethod(method, arguments: arguments)
        }
    }
}

File: android/src/main/kotlin/com/eclypses/socketx_client/SocketXClientPlugin.kt
// The MIT License (MIT)
//
// Copyright (c) Eclypses, Inc.
//
// All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package com.eclypses.socketx_client

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.eclypses.socketx_client_android.SocketXClient
import com.eclypses.socketx_client_android.SocketXError
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString

/**
 * Flutter plugin for secure WebSocket communication using the SocketX protocol.
 */
class SocketXClientPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var methodChannel: MethodChannel
    private var socketXClient: SocketXClient? = null
    private var webSocket: WebSocket? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    // The SocketX library holds no Context of its own, so the plugin supplies a
    // directory for optional log-file capture.
    private var appContext: Context? = null

    companion object {
        private const val CHANNEL_NAME = "socketx_client"
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL_NAME)
        methodChannel.setMethodCallHandler(this)
        appContext = flutterPluginBinding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        disconnect()
        appContext = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "connect" -> handleConnect(call, result)
            "disconnect" -> handleDisconnect(result)
            "sendText" -> handleSendText(call, result)
            "sendBinary" -> handleSendBinary(call, result)
            "enableFileLogging" -> handleEnableFileLogging(call, result)
            "readLogFile" -> result.success(SocketXClient.readLogFile().ifEmpty { null })
            "clearLogFile" -> {
                SocketXClient.clearLogFile()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    // MARK: - Diagnostics

    private fun handleEnableFileLogging(call: MethodCall, result: Result) {
        val enabled = call.argument<Boolean>("enabled")
        if (enabled == null) {
            result.error("INVALID_ARGUMENTS", "'enabled' is required", null)
            return
        }
        SocketXClient.enableFileLogging(enabled, appContext?.filesDir)
        result.success(null)
    }

    // MARK: - Connection Methods

    private fun handleConnect(call: MethodCall, result: Result) {
        val url = call.argument<String>("url")
        if (url.isNullOrEmpty()) {
            result.error("INVALID_ARGUMENTS", "URL is required", null)
            return
        }

        // Disconnect existing connection if any
        disconnect()

        try {
            // Build OkHttpClient
            val okHttpClient = OkHttpClient.Builder().build()

            // Build request with optional headers
            val requestBuilder = Request.Builder().url(url)
            call.argument<Map<String, String>>("headers")?.forEach { (key, value) ->
                requestBuilder.addHeader(key, value)
            }
            val request = requestBuilder.build()

            // Initialize SocketXClient
            socketXClient = SocketXClient(okHttpClient).apply {
                onGlobalError = { error ->
                    sendError(error)
                }
            }

            // Create WebSocket with listener
            webSocket = socketXClient?.newWebSocket(request, createWebSocketListener())

            result.success(null)
        } catch (e: Exception) {
            sendError("internal", e.message ?: "Unknown error during connection")
            result.error("CONNECTION_ERROR", e.message, null)
        }
    }

    private fun handleDisconnect(result: Result) {
        disconnect()
        result.success(null)
    }

    private fun disconnect() {
        webSocket?.close(1000, "Client disconnected")
        webSocket = null
        socketXClient = null
    }

    // MARK: - Send Methods

    private fun handleSendText(call: MethodCall, result: Result) {
        val text = call.argument<String>("text")
        if (text.isNullOrEmpty()) {
            result.error("INVALID_ARGUMENTS", "Text is required", null)
            return
        }

        val ws = webSocket
        if (ws == null) {
            result.error("NOT_CONNECTED", "WebSocket is not connected", null)
            return
        }

        ws.send(text)
        result.success(null)
    }

    private fun handleSendBinary(call: MethodCall, result: Result) {
        val data = call.argument<ByteArray>("data")
        if (data == null) {
            result.error("INVALID_ARGUMENTS", "Binary data is required", null)
            return
        }

        val ws = webSocket
        if (ws == null) {
            result.error("NOT_CONNECTED", "WebSocket is not connected", null)
            return
        }

        ws.send(ByteString.of(*data))
        result.success(null)
    }

    // MARK: - WebSocket Listener

    private fun createWebSocketListener(): WebSocketListener {
        return object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                invokeMethod("onConnected", null)
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                invokeMethod("onMessage", text)
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                invokeMethod("onBinaryMessage", bytes.toByteArray())
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                sendError("network", t.message ?: "Connection failed")
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(code, reason)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                // Connection closed normally
            }
        }
    }

    // MARK: - Error Handling

    private fun sendError(error: SocketXError) {
        val (type, reason) = when (error) {
            is SocketXError.NetworkError -> "network" to error.message
            is SocketXError.TransportError -> "transport" to error.message
            is SocketXError.CodecError -> "codec" to error.message
            is SocketXError.HandshakeError -> "handshake" to error.message
            is SocketXError.ProxyError -> "proxy" to error.message
            is SocketXError.InternalError -> "internal" to error.message
            else -> "unknown" to (error.message ?: "Unknown error")
        }
        sendError(type, reason ?: "Unknown error")
    }

    private fun sendError(type: String, reason: String) {
        val errorMap = mapOf(
            "type" to type,
            "reason" to reason
        )
        invokeMethod("onError", errorMap)
    }

    // MARK: - Helper Methods

    private fun invokeMethod(method: String, arguments: Any?) {
        mainHandler.post {
            methodChannel.invokeMethod(method, arguments)
        }
    }
}
